import basic libraries, we import as we go solving the problem

In [2]:
import numpy as np
import pandas as pd

Read the data as it is

In [3]:
df = pd.read_csv('emails.csv')
df.head()

text spam Unnamed: 2  \
0  Subject: naturally irresistible your corporate...    1        NaN   
1  Subject: the stock trading gunslinger  fanny i...    1        NaN   
2  Subject: unbelievable new homes made easy  im ...    1        NaN   
3  Subject: 4 color printing special  request add...    1        NaN   
4  Subject: do not have money , get software cds ...    1        NaN   

  Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8  \
0        NaN        NaN        NaN        NaN        NaN        NaN   
1        NaN        NaN        NaN        NaN        NaN        NaN   
2        NaN        NaN        NaN        NaN        NaN        NaN   
3        NaN        NaN        NaN        NaN        NaN        NaN   
4        NaN        NaN        NaN        NaN        NaN        NaN   

  Unnamed: 9  ... Unnamed: 100 Unnamed: 101 Unnamed: 102 Unnamed: 103  \
0        NaN  ...          NaN          NaN          NaN          NaN   
1        NaN  ...          NaN          NaN          NaN          NaN   
2        NaN  ...          NaN          NaN          NaN          NaN   
3        NaN  ...          NaN          NaN          NaN          NaN   
4        NaN  ...          NaN          NaN          NaN          NaN   

  Unnamed: 104 Unnamed: 105 Unnamed: 106 Unnamed: 107 Unnamed: 108  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

  Unnamed: 109  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 110 columns]

We found that only first 2 columns have values rest are all nans, hence select only first 2 columns

In [4]:
df = df[['text','spam']]
df.head()

text spam
0  Subject: naturally irresistible your corporate...    1
1  Subject: the stock trading gunslinger  fanny i...    1
2  Subject: unbelievable new homes made easy  im ...    1
3  Subject: 4 color printing special  request add...    1
4  Subject: do not have money , get software cds ...    1

Check for any missing values in rows

In [5]:
df.isna().sum()

text    0
spam    2
dtype: int64

2 values in spam column are missing but find out if they happen to be spam or not spam before imputing or dropping

In [11]:
df.spam.value_counts()

0                                                                                               4359
1                                                                                               1368
 its termination would not  have such a phenomenal impact on the power situation .  however        1
 mr suresh prabhu                                                                                  1
Name: spam, dtype: int64

In [9]:
df.isna().value_counts()

text   spam 
False  False    5729
       True        2
dtype: int64

In [15]:
df[df['spam'].isna()]

text spam
1380  Subject: from the enron india newsdesk - april...  NaN
2651  Subject: from the enron india newsdesk - april...  NaN

Keeping in mind the good distribution of data 0 - 4359, 1- 1368, lets drop the missing values

In [16]:
df.dropna(inplace=True)

In [17]:
df.isna().sum()

text    0
spam    0
dtype: int64

We previously saw the distribution had some text values in spam instead of 0's and 1's, lets get those values.

To get those values, just negate values rows with 0's and 1's 

In [28]:
df[~df['spam'].str.contains('0') & ~df['spam'].str.contains('1')]

text  \
1381  e dpc contributed only 0 . 7 per  cent of the ...   
2652  lf against undeserved claims in the event of e...   

                                                   spam  
1381   its termination would not  have such a phenom...  
2652                                  mr suresh prabhu

Now we need to drop these rows a they are not useful for our model

In [30]:
df.shape

(5729, 2)

In [35]:
df.drop(labels = [1381, 2652], inplace=True)

In [36]:
df[~df['spam'].str.contains('0') & ~df['spam'].str.contains('1')]

Empty DataFrame
Columns: [text, spam]
Index: []

In [37]:
df.spam.value_counts()

0    4359
1    1368
Name: spam, dtype: int64

Now let us drop duplicated rows

In [41]:
df.duplicated().sum()

33

Lets have a closer look at the duplicated rows

In [42]:
df[df.duplicated()]

text spam
2156  Subject: research allocations to egm  hi becky...    0
2261  Subject: departure of grant masson  the resear...    0
2413  Subject: re : schedule and more . .  jinbaek ,...    0
2474  Subject: day off tuesday  stinson ,  i would l...    0
2765  Subject: re : your mail  zhendong ,  dr . kami...    0
3125  Subject: re : grades  pam ,  the students rese...    0
3154  Subject: tiger evals - attachment  tiger hosts...    0
3250  Subject: re : i am zhendong  zhendong ,  thank...    0
3251  Subject: hello from enron  dear dr . mcmullen ...    0
3389  Subject: term paper  dr . kaminski ,  attached...    0
3575  Subject: telephone interview with the enron re...    0
3662  Subject: re : summer work . .  jinbaek ,  this...    0
3692  Subject: re : weather and energy price data  m...    0
3825  Subject: research get - together at sandeep ko...    0
4205  Subject: re : willow and pathstar evaluations ...    0
4392  Subject: re : eprm 2001 houston  layla ,  my a...    0
4396  Subject: vp & director count for the research ...    0
4773  Subject: re : weather and energy price data  m...    0
4787  Subject: re : interviews  vince ,  no problem ...    0
5075  Subject: re : grades  pam ,  another group :  ...    0
5182  Subject: phone time  dear dr . kaminski  thank...    0
5218  Subject: re : term project :  brian ,  no prob...    0
5318  Subject: re :  frank ,  yes .  vince  from : f...    0
5342  Subject: re : enron visit - - thanks  larry , ...    0
5475  Subject: re : pserc industrial advisory board ...    0
5523  Subject: re : get together this coming tuesday...    0
5544  Subject: re : contact info  glenn ,  please , ...    0
5630  Subject: retail markets conference  i would li...    0
5634  Subject: term project :  this is the list of p...    0
5666  Subject: june 21 - 22 retail electricity confe...    0
5676  Subject: re : enron weather research  good aft...    0
5700  Subject: schedule and more . .  dr . kaminski ...    0
5718  Subject: * special notification * aurora versi...    0

In [43]:
df[df.duplicated()]['text']

2156    Subject: research allocations to egm  hi becky...
2261    Subject: departure of grant masson  the resear...
2413    Subject: re : schedule and more . .  jinbaek ,...
2474    Subject: day off tuesday  stinson ,  i would l...
2765    Subject: re : your mail  zhendong ,  dr . kami...
3125    Subject: re : grades  pam ,  the students rese...
3154    Subject: tiger evals - attachment  tiger hosts...
3250    Subject: re : i am zhendong  zhendong ,  thank...
3251    Subject: hello from enron  dear dr . mcmullen ...
3389    Subject: term paper  dr . kaminski ,  attached...
3575    Subject: telephone interview with the enron re...
3662    Subject: re : summer work . .  jinbaek ,  this...
3692    Subject: re : weather and energy price data  m...
3825    Subject: research get - together at sandeep ko...
4205    Subject: re : willow and pathstar evaluations ...
4392    Subject: re : eprm 2001 houston  layla ,  my a...
4396    Subject: vp & director count for the research ...
4773    Subjec

In [47]:
df['text'].nunique()

5694

In [49]:
df['spam'].value_counts().sum()

5727

In [50]:
5727-5694

33

Let us drop these duplicate rows

In [51]:
df = df.drop_duplicates()

In [52]:
df.shape

(5694, 2)

In [53]:
df['spam'].value_counts().sum()

5694

In [54]:
df.head(10)

text spam
0  Subject: naturally irresistible your corporate...    1
1  Subject: the stock trading gunslinger  fanny i...    1
2  Subject: unbelievable new homes made easy  im ...    1
3  Subject: 4 color printing special  request add...    1
4  Subject: do not have money , get software cds ...    1
5  Subject: great nnews  hello , welcome to medzo...    1
6  Subject: here ' s a hot play in motion  homela...    1
7  Subject: save your money buy getting this thin...    1
8  Subject: undeliverable : home based business f...    1
9  Subject: save your money buy getting this thin...    1

Now lets do text-preprocessing

In [55]:
df.iloc[30]['text']

'Subject: want to accept credit cards ? 126432211  aredit cpproved  no cecks  do it now  126432211'

As we see text column has values with special characters, lets write a function to remove them first

In [56]:
def remove_special_chars(row):
    
    result = '' #have a predefined empty string, later on can be concatenated to
    
    for i in row: #In each row
        if i.isalnum() or i == ' ': #if the character is alphanumericor a space,
            result = result + i  #retain the character
        else:                      #otherwise
            result = result + ''  #just replace it with empty string
    return result.strip()

In [57]:
df['text'].head(10)

0    Subject: naturally irresistible your corporate...
1    Subject: the stock trading gunslinger  fanny i...
2    Subject: unbelievable new homes made easy  im ...
3    Subject: 4 color printing special  request add...
4    Subject: do not have money , get software cds ...
5    Subject: great nnews  hello , welcome to medzo...
6    Subject: here ' s a hot play in motion  homela...
7    Subject: save your money buy getting this thin...
8    Subject: undeliverable : home based business f...
9    Subject: save your money buy getting this thin...
Name: text, dtype: object

In [60]:
df['text'].apply(lambda x: x.lower())

0       subject naturally irresistible your corporate ...
1       subject the stock trading gunslinger  fanny is...
2       subject unbelievable new homes made easy  im w...
3       subject 4 color printing special  request addi...
4       subject do not have money  get software cds fr...
                              ...                        
5726    subject re  receipts from visit  jim   thanks ...
5727    subject re  enron case study update  wow  all ...
5728    subject re  interest  david   please  call shi...
5729    subject news  aurora 5  2 update  aurora versi...
5730    subjecet  congratulation hi harsh you have bee...
Name: text, Length: 5694, dtype: object

In [61]:
df['text'].head(10)

0    subject naturally irresistible your corporate ...
1    subject the stock trading gunslinger  fanny is...
2    subject unbelievable new homes made easy  im w...
3    subject 4 color printing special  request addi...
4    subject do not have money  get software cds fr...
5    subject great nnews  hello  welcome to medzonl...
6    subject here  s a hot play in motion  homeland...
7    subject save your money buy getting this thing...
8    subject undeliverable  home based business for...
9    subject save your money buy getting this thing...
Name: text, dtype: object

In [62]:
df['text'].apply(remove_special_chars)

0       subject naturally irresistible your corporate ...
1       subject the stock trading gunslinger  fanny is...
2       subject unbelievable new homes made easy  im w...
3       subject 4 color printing special  request addi...
4       subject do not have money  get software cds fr...
                              ...                        
5726    subject re  receipts from visit  jim   thanks ...
5727    subject re  enron case study update  wow  all ...
5728    subject re  interest  david   please  call shi...
5729    subject news  aurora 5  2 update  aurora versi...
5730    subjecet  congratulation hi harsh you have bee...
Name: text, Length: 5694, dtype: object

In [63]:
df['text'].head(10)

0    subject naturally irresistible your corporate ...
1    subject the stock trading gunslinger  fanny is...
2    subject unbelievable new homes made easy  im w...
3    subject 4 color printing special  request addi...
4    subject do not have money  get software cds fr...
5    subject great nnews  hello  welcome to medzonl...
6    subject here  s a hot play in motion  homeland...
7    subject save your money buy getting this thing...
8    subject undeliverable  home based business for...
9    subject save your money buy getting this thing...
Name: text, dtype: object

Now that we have made entire row as lowercase as well as removed all special chars, we can take up some nlp tasks

In [64]:
import sklearn
from sklearn.feature_extraction.text import CountVectorizer

In [65]:
cv = CountVectorizer(stop_words='english',max_features=10000)

In [66]:
X = cv.fit_transform(df['text']).toarray()

In [69]:
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [70]:
X.shape

(5694, 10000)

In [72]:
df['spam'].astype('int')

0       1
1       1
2       1
3       1
4       1
       ..
5726    0
5727    0
5728    0
5729    0
5730    0
Name: spam, Length: 5694, dtype: int32

In [73]:
df['spam'].dtype

dtype('int32')

In [74]:
y = df['spam'].values

In [75]:
y

array([1, 1, 1, ..., 0, 0, 0])

In [76]:
y.shape

(5694,)

In [77]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2)

In [78]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(4555, 10000)
(1139, 10000)
(4555,)
(1139,)


In [79]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score

In [80]:
clf = MultinomialNB()

In [81]:
clf.fit(X_train,y_train)

MultinomialNB()

In [82]:
y_pred = clf.predict(X_test)

In [83]:
accuracy_score(y_test,y_pred)

0.9929762949956101

In [84]:
from sklearn.metrics import classification_report

In [86]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00       842
           1       0.98      0.99      0.99       297

    accuracy                           0.99      1139
   macro avg       0.99      0.99      0.99      1139
weighted avg       0.99      0.99      0.99      1139



In [88]:
import nltk
from nltk.stem.porter import PorterStemmer

ps=PorterStemmer()

df['text'] = df['text'].apply(lambda x: [ps.stem(y) for y in x])

In [89]:
def joinList(text):
    return ''.join(text)

In [90]:
df['text']=df['text'].apply(joinList)

In [91]:
df.sample(3)

text  spam
4513  subject re  re  conference volume  hi vince   ...     0
3290  subject reliant tigers interested in enron  vi...     0
3134  subject lsu visit  resume   mr  kaminski   it ...     0

In [92]:
X1=cv.fit_transform(df['text']).toarray()
y1=df.iloc[:,-1].values

In [93]:
X1_train,X1_test,y1_train,y1_test=train_test_split(X1,y1,test_size=0.2,random_state=0)

In [94]:
clf2 = MultinomialNB()
clf2.fit(X1_train,y1_train)
y1_pred=clf2.predict(X1_test)

In [95]:
print(classification_report(y1_test, y1_pred))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99       874
           1       0.97      0.97      0.97       265

    accuracy                           0.99      1139
   macro avg       0.98      0.98      0.98      1139
weighted avg       0.99      0.99      0.99      1139



In [96]:
print(accuracy_score(y1_test, y1_pred))

0.9868305531167691
